internal/cmd/launcher.go

```diff
@@ -0,0 +1,348 @@
+
+	rc := 0
+	if err := cmd.Run(); err != nil {
+		rc = 1
+		gs.Logger.
+			WithError(err).
+			Error("Failed to run the provisioned k6 binary")
+
+		var eerr *exec.ExitError
+		if errors.As(err, &eerr) {
+			rc = eerr.ExitCode()
+		}
+	}
+	gs.OSExit(rc)
+}
+
+// currentBinary runs the requested commands on the current binary
+type currentBinary struct{}
+
+func (b *currentBinary) run(gs *state.GlobalState) {
+	ExecuteWithGlobalState(gs)
+}
+
+// customBuildRequired checks if the build is required
+// it's required if there is one or more dependencies other than k6 itself
+// or if the required k6 version is not satisfied by the current binary's version
+// TODO: get the version of any built-in extension and check if they satisfy the dependencies
+func customBuildRequired(baseK6Version string, deps k6deps.Dependencies) bool {
+	if len(deps) == 0 {
+		return false
+	}
+
+	// Early return if there are multiple dependencies
+	if len(deps) > 1 {
+		return true
+	}
+
+	k6Dependency, hasK6 := deps["k6"]
+
+	// Early return if there's exactly one non-k6 dependency
+	if !hasK6 {
+		return true
+	}
+
+	// Ignore k6 dependency if nil
+	if k6Dependency == nil || k6Dependency.Constraints == nil {
+		return false
+	}
+
+	k6Ver, err := semver.NewVersion(baseK6Version)
+	if err != nil {
+		// ignore if baseK6Version is not a valid sem ver (e.g. a development version)
+		return true
+	}
+
+	// if the current version satisfies the constrains, binary provisioning is not required
+	return !k6Dependency.Constraints.Check(k6Ver)
+}
+
+// k6buildProvision returns the path to a k6 binary that satisfies the dependencies and the list of versions it provides
+func k6buildProvision(gs *state.GlobalState, deps k6deps.Dependencies) (commandExecutor, error) {
+	token, err := extractToken(gs)
+	if err != nil {
+		gs.Logger.WithError(err).Debug("Failed to get a valid token")
+	}
+
+	if token == "" {
+		return nil, errors.New("k6 cloud token is required when the Binary provisioning feature is enabled." +
+			" Set K6_CLOUD_TOKEN environment variable or execute the `k6 cloud login` command")
+	}
+
+	config := k6provider.Config{
+		BuildServiceURL:  gs.Flags.BuildServiceURL,
+		BuildServiceAuth: token,
+	}
+
+	provider, err := k6provider.NewProvider(config)
+	if err != nil {
+		return nil, err
+	}
+
+	binary, err := provider.GetBinary(gs.Ctx, deps)
+	if err != nil {
+		return nil, err
+	}
+
+	gs.Logger.
+		Info("A new k6 binary has been provisioned with version(s): ", formatDependencies(binary.Dependencies))
+
+	return &customBinary{binary.Path}, nil
+}
+
+func formatDependencies(deps map[string]string) string {
+	buffer := &bytes.Buffer{}
+	for dep, version := range deps {
+		buffer.WriteString(fmt.Sprintf("%s:%s ", dep, version))
+	}
+	return strings.Trim(buffer.String(), " ")
+}
+
+// extractToken gets the cloud token required to access the build service
+// from the environment or from the config file
+func extractToken(gs *state.GlobalState) (string, error) {
```

Comment:
Is it intentional that while most (if not all) of the other Cloud commands can still read old configuration, this one can't, and just tells you that you don't have a valid token?

If this is a way to force users to move to the new configuration if they want to test the feature, then it's alright. 

But, if not, then we should fix it, because in my opinion is a bit confusing: I can successfully run a Cloud command, like `k6 cloud run`, but when I run it with binary provisioning it tells me that I don't have a valid token, while both rely on the same token.

Reply:
>If this is a way to force users to move to the new configuration if they want to test the feature, then it's alright. It may cause some friction and confusion, but it's the "cost" of forcing users to move forward.

@joanlopez Yes, this was my intention. I wanted to reduce the already complex world around this feature. I guess it's an acceptable scenario. Let me know if you disagree.

Reply:
If we all agree about the implications, it's fine! I see no major problems beyond what's obvious and aforementioned.

Reply:
I don't have a strong opinion here. `k6exec` supported old configurations. I was expecting that by reusing `k6` logic we would inherit this ability. But I'm ok with not supporting it. Should we add an item to follow-up this or we consider it settled?